docs/api-communication-spec.md

```diff
@@ -0,0 +1,686 @@
+- **BOSH** Deployment VMs, where network traffic is captured on the VM.
+
+Pcap-release uses gRPC for communication between all components. This document highlights the intended use of fields, scenarios considered by the implementation and detailed explanation of the data structures.
+
+## CF Case Overview
+
+Capturing data from cloud foundry applications requires the pcap-agent inside the app container. The pcap-agent runs in plain-text HTTP/2 (H2C). The Envoy in the app container provides the TLS termination.
+
+The Cloud Controller is used to identify the target ports of the pcap-agent in the target app instance.
+
+```mermaid
+graph LR
+    pcap-cli -->|H/2, gRPC| gorouter
+    gorouter -->|H/2, gRPC| pcap-api
+
+    subgraph pcap deployment
+        pcap-api
+    end
+    
+    subgraph diego-cell1
+        subgraph cf-app1
+            pcap-api -->|H/2, gRPC| c1e[Envoy]
+            c1e -->|gRPC, H2C| c1[pcap-agent]
+        end 
+        subgraph cf-app2
+            pcap-api -->|H/2, gRPC| c2e[Envoy]
+            c2e -->|gRPC, H2C| c2[pcap-agent]
+        end 
+    end
+    subgraph diego-cell2
+        subgraph cf-app3
+            pcap-api -->|H/2, gRPC| c3e[Envoy]
+            c3e -->|H2C, gRPC| c3[pcap-agent]
+        end 
+    end
+    subgraph cloud-controller
+        pcap-api -->|HTTP/1.1, JSON| cc[Cloud Controller]
+    end
+```
+Figure 1: The Cloud Foundry Application Capturing case
+
+## BOSH Case Overview
+
+Capturing data from BOSH deployments requires the pcap-agent on the individual BOSH VMs. The pcap-agent runs on a well-known port and takes care of mTLS verification.
+
+The BOSH director is used to verify and find the requested target VMs' IP addresses. The BOSH UAA is used to verify the token provided by the client.
+
+```mermaid
+graph LR
+    pcap-cli -->|H/2, gRPC| pcap-api
+
+    subgraph pcap deployment
+        pcap-api
+    end
+        
+    subgraph bosh-deployment-vm1
+        pcap-api -->|H/2, gRPC| a1[pcap-agent]
+    end
+    subgraph bosh-deployment-vm2
+        pcap-api -->|H/2, gRPC| a2[pcap-agent]
+    end
+
+    subgraph bosh-director-vm
+        pcap-api -->|HTTP/1.1, JSON| bosh[BOSH Director]
+        pcap-api -->|HTTP/1.1, JSON| bosh-uaa[BOSH UAA]
+    end
+```
+
+Figure 2: The BOSH Capturing case
+
+## Interactions
+
+### pcap-cli -> pcap-api
+  - request status (version information, service alive)
+  - capture request for:
+    - CF Capture Request
+    - BOSH Capture Request
+  - Send stop request
+### pcap-api
+  - Status response
+  - CF Capture Request
+    - pcap-api -> Cloud Controller
+      - Validate request, token
+      - find endpoints (pcap-agent)
+  - BOSH Capture Request
+    - pcap-api -> BOSH Director
+      - HTTP: Retrieve /info (also works without Auth)
+      - HTTP: Retrieve deployment instances
+    - pcap-api -> BOSH Director UAA
+      - HTTP: Retrieve JWT token key
```

Comment:
Are you referring to refreshing of the access token? Because the initial token should be provided by the user.

Reply:
No, the JWT token for BOSH is signed with an RSA key. The public key is among the "JWT Token key" information retrieved from BOSH UAA (endpoint `/token_keys`).

We did not cover the case of pcap-bosh-cli's use case of token refresh though. so we will add that.